aws/credential_cache_test.go

```diff
@@ -211,6 +211,121 @@ func TestCredentialsCache_ExpireTime(t *testing.T) {
 	}
 }
 
+func TestCredentialsCache_AsyncRefresh(t *testing.T) {
+	orig := sdk.NowTime
+	defer func() { sdk.NowTime = orig }()
+	var mockTime time.Time
+	sdk.NowTime = func() time.Time { return mockTime }
+
+	cases := []struct {
+		Creds  func() Credentials
+		Called int32
+	}{
+		{
+			Called: 6,
+			Creds: func() Credentials {
+				return Credentials{
+					AccessKeyID:     "key",
+					SecretAccessKey: "secret",
+					CanExpire:       true,
+					Expires:         mockTime.Add(5),
+				}
+			},
+		},
+	}
+
+	for _, c := range cases {
+		var called int32
+		p := NewCredentialsCache(CredentialsProviderFunc(func(ctx context.Context) (Credentials, error) {
+			atomic.AddInt32(&called, 1)
+			return c.Creds(), nil
+		}), func(options *CredentialsCacheOptions) {
+			options.EnableAsyncRefresh = true
+		})
+
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+
+		mockTime = mockTime.Add(10)
+
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+		p.Retrieve(context.Background())
+		testWaitAsyncRefreshDone(p)
+
+		if e, a := c.Called, called; e != a {
+			t.Errorf("expect %v called, got %v", e, a)
+		}
+	}
+}
+
+func TestCredentialsCache_AsyncRefreshWithMinimumDelay(t *testing.T) {
+	orig := sdk.NowTime
+	defer func() { sdk.NowTime = orig }()
+	var mockTime time.Time
+	sdk.NowTime = func() time.Time { return mockTime }
+
+	cases := []struct {
+		Creds  func() Credentials
+		Called int32
+	}{
+		{
+			Called: 4,
+			Creds: func() Credentials {
+				return Credentials{
+					AccessKeyID:     "key",
+					SecretAccessKey: "secret",
+					CanExpire:       true,
+					Expires:         mockTime.Add(5),
+				}
+			},
+		},
+	}
+
+	for _, c := range cases {
+		var called int32
+		p := NewCredentialsCache(CredentialsProviderFunc(func(ctx context.Context) (Credentials, error) {
+			atomic.AddInt32(&called, 1)
+			return c.Creds(), nil
+		}), func(options *CredentialsCacheOptions) {
+			options.EnableAsyncRefresh = true
+			options.AsyncRefreshMinimumDelay = 2 * time.Nanosecond
+		})
+
+		p.Retrieve(context.Background())
```

Comment:
can we add some comments in here to describe _why_ the expected result is what it is, e.g.

```go
// first call - no async refresh
p.Retrieve(...)
testWaitAsyncRefreshDone(p)

// cached after first call, async refresh
p.Retrieve(...)
testWaitAsyncRefreshDone(p)
// within minimum delay, no refresh should happen
...

mockTime = mockTime.Add(3)

// not expired but outside the delay
p.Retrieve(...)
testWaitAsyncRefreshDone(p)

...

mockTime = mockTime.add(1)
```

This is just an example, the behavior of this test isn't quite right anyway though, see my comment above on the behavior not matching the docs.